addons/account/tests/test_account_move_reconcile.py

```diff
@@ -4000,6 +4000,90 @@ def test_reconcile_cash_basis_tax_grid_multi_taxes(self):
+        line_d = self.create_line_for_reconciliation(300, 300, currency, '2016-01-01')
+        (line_a + line_b).reconcile()
+        (line_b + line_c).reconcile()
+        (line_c + line_d).reconcile()
+        self.assertEqual(line_a.matching_number, f'P{line_a.matched_debit_ids.ids[0]}')
+        self.assertEqual(line_a.matching_number, line_b.matching_number)
+        self.assertEqual(line_a.matching_number, line_c.matching_number)
+        self.assertEqual(line_a.matching_number, line_d.matching_number)
+
+    def test_matching_number_partial_multi_separate_reconcile(self):
+        currency = self.currency_data['currency']
+        line_a = self.create_line_for_reconciliation(1000, 1000, currency, '2016-01-01')
+        line_b = self.create_line_for_reconciliation(-500, -500, currency, '2016-01-01')
+        (line_a + line_b).reconcile()
+        self.assertEqual(line_a.matching_number, f'P{line_a.matched_credit_ids.id}')
+        self.assertEqual(line_a.matching_number, line_b.matching_number)
+
+        line_c = self.create_line_for_reconciliation(-300, -300, currency, '2016-01-01')
+        (line_a + line_c).reconcile()
+        self.assertEqual(line_a.matching_number, f'P{line_a.matched_credit_ids.ids[0]}')
+        self.assertEqual(line_a.matching_number, line_b.matching_number)
+        self.assertEqual(line_a.matching_number, line_c.matching_number)
+
+    def test_matching_number_unreconcile_single(self):
+        currency = self.currency_data['currency']
+        full_line_a = self.create_line_for_reconciliation(200, 200, currency, '2016-01-01')
+        full_line_b = self.create_line_for_reconciliation(-200, -200, currency, '2016-01-01')
+        partial_line_a = self.create_line_for_reconciliation(1000, 1000, currency, '2016-01-01')
+        partial_line_b = self.create_line_for_reconciliation(-500, -500, currency, '2016-01-01')
+        (full_line_a + full_line_b).reconcile()
+        (partial_line_a + partial_line_b).reconcile()
+        (full_line_a + full_line_b + partial_line_a + partial_line_b).remove_move_reconcile()
+        self.assertFalse(full_line_a.matching_number)
+        self.assertFalse(full_line_b.matching_number)
+        self.assertFalse(partial_line_a.matching_number)
+        self.assertFalse(partial_line_b.matching_number)
+
+    def test_matching_number_unreconcile_multi(self):
+        currency = self.currency_data['currency']
+        line_a = self.create_line_for_reconciliation(-500, -500, currency, '2016-01-01')
+        line_b = self.create_line_for_reconciliation(1000, 1000, currency, '2016-01-01')
+        line_c = self.create_line_for_reconciliation(-1000, -1000, currency, '2016-01-01')
+        line_d = self.create_line_for_reconciliation(300, 300, currency, '2016-01-01')
+        (line_a + line_b + line_c + line_d).reconcile()
+
+        previous_matching_number = line_a.matching_number
+        line_d.remove_move_reconcile()
```

Comment:
From the way you reconcile, you can't really know for sure if `line_d` is matched with `line_a` or `line_c`.
Because of this, `line_a` could maybe keep the previous number, or am I missing something?

I think that my suggestion for the test above is more robust, if I'm not missing a case.

Reply:
Yep you're right. I just rewrote them to make it more clear